frontend/test/resolution/testInitSemantics.cpp

```diff
@@ -1737,6 +1737,162 @@ static void testInitGenericAfterConcrete() {
   }
 }
 
+static std::string toString(QualifiedType type) {
+  std::stringstream ss;
+  type.type()->stringify(ss, chpl::StringifyKind::CHPL_SYNTAX);
+  return ss.str();
+}
+
+static void testNilFieldInit() {
+  std::string program = R"""(
+    class C { var x: int; }
+    record R {
+      var x: unmanaged C?;
+      proc init() {
+        x = nil;
+      }
+    }
+
+    // exists to work around current lack of default-init at module scope
+    proc test() {
+      var x: R;
+      return x;
+    }
+    var a = test();
+    var b = new R();
+  )""";
+  auto config = getConfigWithHome();
+  Context ctx(config);
+  Context* context = &ctx;
+  setupModuleSearchPaths(context, false, false, {}, {});
+  ErrorGuard guard(context);
+
+  auto vars = resolveTypesOfVariables(context, program, {"a", "b"});
+
+  assert(toString(vars["a"]) == "R");
+  assert(toString(vars["b"]) == "R");
+}
+
+static void testGenericFieldInit() {
+  {
+    std::string program = R"""(
+      record R {
+        var x : integral;
+
+        proc init(arg) {
+          this.x = arg;
+        }
+      }
+
+      var a = new R(5);
+      var b = new R(10:uint);
+      var c = new R("test");
+      )""";
+
+    auto config = getConfigWithHome();
+    Context ctx(config);
+    Context* context = &ctx;
+    setupModuleSearchPaths(context, false, false, {}, {});
+    ErrorGuard guard(context);
+
+    auto vars = resolveTypesOfVariables(context, program, {"a", "b", "c"});
+
+    assert(guard.numErrors() == 1);
+    auto& err = guard.error(0);
+    assert(err->type() == ErrorType::IncompatibleTypeAndInit);
+    assert(err->location(context).firstLine() == 6);
+
+    // Note: These type strings are not stabilized
+    assert(toString(vars["a"]) == "R(var int(64))");
+    assert(toString(vars["b"]) == "R(var uint(64))");
+    assert(toString(vars["c"]) == "R(var ErroneousType)");
+
+    guard.realizeErrors();
+  }
+  {
+    std::string program = R"""(
+      class C {
+        type typeField;
+      }
+
+      record R {
+        var myC: owned C(?)?;
+
+        proc init() {
+          this.myC = nil;
+        }
+      }
+
+      var r  = new R();
+      )""";
+    auto config = getConfigWithHome();
+    Context ctx(config);
+    Context* context = &ctx;
+    setupModuleSearchPaths(context, false, false, {}, {});
+    ErrorGuard guard(context);
+
+    auto vars = resolveTypesOfVariables(context, program, {"r"});
+    assert(toString(vars["r"]) == "R(var ErroneousType)");
+
+    assert(guard.numErrors() == 1);
+    auto& err = guard.error(0);
+    assert(err->type() == ErrorType::IncompatibleTypeAndInit);
```

Comment:
This is a surprising error to issue here, since `nil` `owned C? = nil` is definitely compatible. I would expect something else.

That said, I'm glad an error is emitted in this case at the right place :)